distributed/shuffle/_worker_extension.py

```diff
@@ -370,9 +367,17 @@ def _() -> dict[str, list[tuple[ArrayRechunkShardID, bytes]]]:
             write the serialized payload directly to disk on the receiver.
             """
             out: dict[str, list[tuple[ArrayRechunkShardID, bytes]]] = defaultdict(list)
-            for id, nslice in self._slicing[input_partition]:
-                out[self.worker_for[id.chunk_index]].append(
-                    (id, pickle.dumps((id.shard_index, data[nslice])))
+            from itertools import product
+
+            shards = product(
+                *(axis[i] for axis, i in zip(self.split_axes, input_partition))
```

Comment:
Is it obvious why the input partition is the same length as the split axes?

Reply:
I think almost none about the rechunking code is obvious, so please raise wherever you're missing information. It took my way too long to understand what's going on in task-bask rechunking. 

`split_axes` contains the splits across all N dimension/axes and input_partition is an N-dimensional index.

Reply:
I've renamed `input_partition` to `partition_id` in case that makes things clearer.